src/protocol/dot_product.rs

```diff
@@ -0,0 +1,173 @@
+///
+/// This protocol is an extension of the CHIKP multiplication protocol described in the paper:
+/// "High-Throughput Secure AES Computation" (High-throughput secure AES computation. In WAHC@CCS 2018, pp. 13–24, 2018)
+/// by Koji Chida, Koki Hamada, Dai Ikarashi, Ryo Kikuchi and Benny Pinkas
+/// <https://dl.acm.org/doi/10.1145/3267973.3267977>
+///
+/// The extension allows for the computation of the sum of multiple multiplications (in other words, a dot-product operation)
+/// at the communication cost of a single multiplication.
+///
+/// This extension is summarized in the paper:
+/// "Fast Large-Scale Honest-Majority MPC for Malicious Adversaries"
+/// by Koji Chida, Daniel Genkin, Koki Hamada, Dai Ikarashi, Ryo Kikuchi, Yehuda Lindell, and Ariel Nof
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// In that paper, this functionality is called `F_product` and
+/// an instantiation for Replciated Secret Sharing is provided in section 6.1 on page 27
+///
+/// To summarize the CHIKP protocol:
+/// each helper computes `u_i+1 = (s_i + s_i+1) * (t_i + t_i+1) - s_i+1 * t_i+1`
+/// It then randomizes this share with randomness that it shares with the
+/// helper to its left, and sends it to the helper to its right,
+/// and receives `u_i` from the helper to its left.
+/// Each helper then defines the pair `(u_i, u_i+1)` as its shares of the output.
+///
+/// (Note it is equivalent to compute `u_i+1 = s_i * t_i + s_i+1 * t_i + s_i * t_i+1`,
+/// but this requires three multiplications instead of two. Assuming multiplying field elements is
+/// significantly more costly (in CPU terms) than addition, this is less efficient to compute)
+///
+/// The key observation that makes this protocol work, is that each helper can compute many `u_i+1` values
+/// from many multiplications, sum them all together, randomize them with correlated randomness, and can
+/// send this single field value to the next helper. The result will be a replicated secret sharing of the
+/// dot-product of the two vectors.
+///
+pub mod stream {
+    use crate::{
+        error::BoxError,
+        field::Field,
+        helpers::{fabric::Network, Direction},
+        protocol::{context::ProtocolContext, RecordId},
+        secret_sharing::Replicated,
+    };
+    use futures::stream::StreamExt; // for `next`
+    use futures::Stream;
+    use std::pin::Pin;
+
+    use serde::{Deserialize, Serialize};
+
+    /// A message sent by each helper when they've computed one share of the result
+    #[derive(Debug, Serialize, Deserialize, Default)]
+    pub struct UValue<F> {
+        u: F,
+    }
+
+    type PairOfShares<F> = (Replicated<F>, Replicated<F>);
```

Comment:
it may be worth creating a wrapper here that explicitly defines which share is $z_k$ and which is $r \times z_k$. 
Looks like if I reorder these shares by sending $z_{k_1}$, $r \times z_{k_1}$ and then $r \times z_{k_2}$  $z_{k_2}$ that will lead to incorrect result so it is worth trying to prevent that

Reply:
Oh, this will only be for just one. We will have two dot-product accumulator string thingies. One for the dot product:
$[u] = \sum [α_k] * [r*z_k]$
and another one for the dot product:
$r[w] = r\sum [α_k] * [z_k]$?

For the first stream, we will pass in pairs of $([α_k], [r*z_k])$.
For the second stream, we will pass in pairs of $([α_k], [z_k])$

And actually, the order will not matter at all because dot_product(a, b) == dot_product(b, a)